File: Hungarian.php
<?php 

namespace iqbalhs\hungarian;

final class Hungarian 
{
    public $matrix = [];

    protected $reduced = [];

    function __construct(array $matrix) {
        $this->validate($matrix);
        $this->reduced = $this->matrix = $matrix;
    }

    protected function validate(array $matrix)
    {
        if (!is_array($matrix)) {
            throw new \Exception('Invalid format for matrix');
        }
        return true;
    }

    public function reduce()
    {
        foreach ($this->reduced as $row => $cells) {
            $min = min($cells);
            foreach ($cells as $column => $cell) {
                $this->reduced[$row][$column] -= $min;
            }
        }
        $transposed = array_map(null, ...$this->reduced);
        foreach ($this->reduced as $column => $cells) {
            $min = min($cells);
            foreach ($cells as $row => $cell) {
                $this->reduced[$row][$column] -= $min;
            }
        }
    }

    public function getReduced()
    {
        return $this->reduced;
    }

    public function print()
    {
        foreach ($this->matrix as $data) {
            print(implode(' ,', $data) . "\n");
        }
    }

    public function solve()
    {
        $this->reduce();
        // TODO continue to next step after reduced it
    }
}

File: hungarian.py
class Hungarian():
    def __init__(self, matrix):
        self.matrix = matrix
        self.reduced = matrix
        self.validate()

    def validate(self):
        if not isinstance(self.matrix, list):
            raise ValueError("Matrix is not valid!")

        return True

    def reduce(self):
        for cells in self.matrix:
            row = self.matrix.index(cells)
            minVal = min(cells)
            for cell in cells:
                column = cells.index(cell)
                self.reduced[row][column] -= minVal
        transposed = [list(x) for x in zip(*self.matrix)]
        for cells in transposed:
            column = transposed.index(cells)
            minVal = min(cells)
            for cell in cells:
                row = cells.index(cell)
                self.reduced[row][column] -= minVal

    def solve(self):
        self.reduce()
